gbuiltins/bls381/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+// This file is part of Gear.
+
+// Copyright (C) Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(not(feature = "std"))]
+extern crate alloc;
+
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+use codec::{Decode, Encode};
+
+/// Constant defines codec index of [`Request::MultiMillerLoop`].
+pub const REQUEST_MULTI_MILLER_LOOP: u8 = 0;
```

Comment:
Why do you need those constants, especially in public?

Reply:
if absolutely required to get hint about payload, put enum + const generation under `macro_rules!` to avoid duplication

Reply:
We need to decode payload manually in builtin actor for precise gas counting since there is no instrumentation similar to gear programs.

Good point for macro, thanks! However enum variants can't be generated by the Rust macro so to implement such kind of logic one has to use push-down accumulation, incremental TT munchers and other approaches. So I'm afraid the macro will be over complicated and take more lines of code than the current implementation.

Therefore if you don't mind I would stay with that solution of defining constants/variants by hands.